File: marketing/emailcreationskill/backend/audience_builder/last_sent.py
"""
Deterministic (non-LLM) "what was sent last time" lookup for the Audience
Builder tab. Reuses the same /marketing/v3/emails list-search endpoint as
hubspot_tools.lookup_brand_history. The include/exclude list IDs live under
each email's `to.contactLists` (static/manual lists) and `to.contactIlsLists`
(active/dynamic lists) — NOT a top-level `sendOptions` field (that field does
not exist on either the list-search or single-email-GET response; confirmed
against live HubSpot data) — so both are read directly off these list-search
results with no extra per-email GET needed.
"""
import re

import audience_tools
from config import HUBSPOT_PORTAL_ID

_STOPWORDS = {"the", "a", "an", "and", "of", "for", "in", "on", "to"}
_YEAR_RE = re.compile(r"\b(19|20)\d{2}\b")


def _keywords(text: str) -> set[str]:
    return {w for w in re.findall(r"[a-z0-9]+", (text or "").lower()) if len(w) > 2 and w not in _STOPWORDS}


def _strip_year(text: str) -> str:
    """Drop a trailing/embedded 4-digit year, e.g. 'MCP Dev Summit Seoul 2026' ->
    'MCP Dev Summit Seoul'. The discovery agent often appends the event's year
    to event_name, which breaks HubSpot's name__icontains search since most
    email names for an event don't repeat the year literally."""
    stripped = _YEAR_RE.sub("", text or "")
    return re.sub(r"\s+", " ", stripped).strip()


def _resolve_list_brief(list_id, cache: dict) -> dict:
    list_id = str(list_id)
    if list_id in cache:
        return cache[list_id]
    try:
        raw = audience_tools.hubspot_get_list(list_id)
        info = raw.get("list", raw)
        brief = {"list_id": list_id, "name": info.get("name") or list_id, "size": info.get("size")}
    except Exception:
        brief = {"list_id": list_id, "name": list_id, "size": None}
    cache[list_id] = brief
    return brief


def find_last_sent_emails(event_name: str = "", brand_short: str = "", limit: int = 3) -> list[dict]:
    """Up to `limit` most-recently-published marketing emails whose name
    references this event, sorted by publishDate desc, with each email's
    included/suppression list IDs resolved to {list_id, name, size}.

    Searches by event_name first (server-side name__icontains, same pattern
    as lookup_brand_history), falling back to brand_short if that comes back
    empty — event names are often abbreviated differently across editions, so
    results are additionally ranked by keyword overlap rather than trusted
    on substring match alone."""
    term = (event_name or brand_short or "").strip()
    if not term:
        return []

    search_term = _strip_year(term) or term
    raw = audience_tools.hubspot_search_emails_raw(search_term, limit=30)
    published = [e for e in raw if e.get("state") == "PUBLISHED"]

    # Retry with brand_short whenever the primary term yields no *published*
    # candidate, not only when it yields none at all — a term can still match
    # a stray AUTOMATED/DRAFT email (e.g. a travel-fund approval notice) while
    # missing every real campaign send.
    if not published and brand_short and brand_short.strip().lower() != search_term.lower():
        raw = audience_tools.hubspot_search_emails_raw(brand_short.strip(), limit=30)
        published = [e for e in raw if e.get("state") == "PUBLISHED"]

    query_kw = _keywords(event_name) | _keywords(brand_short)
    if query_kw:
        scored = [(len(query_kw & _keywords(e.get("name", ""))), e) for e in published]
        matched = [pair for pair in scored if pair[0] > 0]
        candidates = matched if matched else [(0, e) for e in published]
    else:
        candidates = [(0, e) for e in published]

    # hubspot_search_emails_raw already sorts -publishDate server-side;
    # a stable sort on keyword score alone preserves that as the tiebreak.
    candidates.sort(key=lambda pair: pair[0], reverse=True)

    cache: dict = {}
    results = []
    for _, e in candidates[:limit]:
        to = e.get("to") or {}
        static_lists = to.get("contactLists") or {}
        active_lists = to.get("contactIlsLists") or {}
        included_ids = list(static_lists.get("include") or []) + list(active_lists.get("include") or [])
        excluded_ids = list(static_lists.get("exclude") or []) + list(active_lists.get("exclude") or [])
        email_id = e.get("id")
        results.append({
            "email_id": email_id,
            "email_name": e.get("name"),
            "sent_at": e.get("publishDate") or e.get("updatedAt"),
            "hubspot_url": (
                f"https://app.hubspot.com/email/{HUBSPOT_PORTAL_ID}/edit/{email_id}"
                if email_id else ""
            ),
            "included_lists": [_resolve_list_brief(lid, cache) for lid in included_ids],
            "suppression_lists": [_resolve_list_brief(lid, cache) for lid in excluded_ids],
        })
    return results

File: marketing/email-creation-service/backend/internal/service/audience/last_sent.go
package audience

import (
	"context"
	"errors"
	"fmt"
	"regexp"
	"sort"
	"strings"

	"github.com/linuxfoundation/lfx-v2-emailcreation-service/internal/domain"
	"github.com/linuxfoundation/lfx-v2-emailcreation-service/internal/domain/model"
)

// LastSentService finds previously-sent marketing emails for an event and
// resolves their include/exclude list IDs to display info — ported from
// audience_builder/last_sent.py.
type LastSentService struct {
	emails   domain.HubSpotEmailClient
	lists    domain.HubSpotListClient
	portalID string
}

func NewLastSentService(emails domain.HubSpotEmailClient, lists domain.HubSpotListClient, portalID string) *LastSentService {
	return &LastSentService{emails: emails, lists: lists, portalID: portalID}
}

var yearRe = regexp.MustCompile(`\b(19|20)\d{2}\b`)
var whitespaceRe = regexp.MustCompile(`\s+`)

// stripYear drops a trailing/embedded 4-digit year, e.g. "MCP Dev Summit
// Seoul 2026" -> "MCP Dev Summit Seoul". The discovery agent often appends
// the event's year to event_name, which breaks HubSpot's name__icontains
// search since most email names for an event don't repeat the year
// literally.
func stripYear(text string) string {
	stripped := yearRe.ReplaceAllString(text, "")
	return strings.TrimSpace(whitespaceRe.ReplaceAllString(stripped, " "))
}

// resolveListBrief resolves a list ID referenced on a past send to
// {list_id, name, size}. The email's to.contactLists/contactIlsLists fields
// freeze whatever list ID was attached at send time; when that ID 404s on
// v3 it usually isn't gone — it's the old (v1) numbering for a list HubSpot
// has since carried forward under a new v3 listId. We recover the live v3
// list via the legacy v1 endpoint (for the name) + a v3 name search (for the
// current ID/size), and only fall back to missing=true if that recovery
// fails too, i.e. the list is actually gone, not just renumbered.
func (s *LastSentService) resolveListBrief(ctx context.Context, listID string, cache map[string]model.ListBrief) model.ListBrief {
	if b, ok := cache[listID]; ok {
		return b
	}

	info, err := s.lists.GetList(ctx, listID)
	if err == nil {
		name := info.Name
		if name == "" {
			name = listID
		}
		brief := model.ListBrief{ListID: listID, Name: name, Size: optionalSize(info.Size)}
		cache[listID] = brief
		return brief
	}
	if !isNotFoundErr(err) {
		brief := model.ListBrief{ListID: listID, Name: fmt.Sprintf("List %s (lookup failed)", listID), Missing: true}
		cache[listID] = brief
		return brief
	}

	legacyName, exists, legacyErr := s.lists.GetLegacyListName(ctx, listID)
	if legacyErr == nil && exists && legacyName != "" {
		if results, searchErr := s.lists.SearchLists(ctx, legacyName, 20); searchErr == nil {
			for _, r := range results {
				if r.Name == legacyName && r.ID != "" {
					brief := model.ListBrief{ListID: r.ID, Name: r.Name, Size: optionalSize(r.Size), ResolvedFromLegacyID: listID}
					cache[listID] = brief
					return brief
				}
			}
		}
		brief := model.ListBrief{
			ListID:  listID,
			Name:    fmt.Sprintf("%q (renumbered in HubSpot; original ID %s could not be re-matched)", legacyName, listID),
			Missing: true,
		}
		cache[listID] = brief
		return brief
	}

	brief := model.ListBrief{ListID: listID, Name: fmt.Sprintf("List %s (no longer exists in HubSpot)", listID), Missing: true}
	cache[listID] = brief
	return brief
}

func isNotFoundErr(err error) bool {
	return err != nil && errors.Is(err, domain.ErrNotFound)
}

// dedupeBriefs collapses briefs referencing the same resolved list_id — a
// raw ID list can contain both a legacy v1 ID and its live v3 ID for the
// same list (e.g. static + dynamic recipient fields both referencing it),
// and resolveListBrief resolves the legacy one to the same v3 list_id.
func dedupeBriefs(briefs []model.ListBrief) []model.ListBrief {
	seen := map[string]bool{}
	out := make([]model.ListBrief, 0, len(briefs))
	for _, b := range briefs {
		if seen[b.ListID] {
			continue
		}
		seen[b.ListID] = true
		out = append(out, b)
	}
	return out
}

type scoredEmail struct {
	score int
	email model.EmailSummary
}

// FindLastSentEmails returns up to `limit` most-recently-published
// marketing emails whose name references this event, sorted by publishDate
// desc, with each email's included/suppression list IDs resolved to
// {list_id, name, size}.
//
// Searches by event_name first (server-side name__icontains), falling back
// to brand_short if that comes back with no published email — event names
// are often abbreviated differently across editions, so results are
// additionally ranked by keyword overlap rather than trusted on substring
// match alone.
func (s *LastSentService) FindLastSentEmails(ctx context.Context, eventName, brandShort string, limit int) ([]model.LastSentEmail, error) {
	if limit <= 0 {
		limit = 3
	}
	term := strings.TrimSpace(eventName)
	if term == "" {
		term = strings.TrimSpace(brandShort)
	}
	if term == "" {
		return []model.LastSentEmail{}, nil
	}

	searchTerm := stripYear(term)
	if searchTerm == "" {
		searchTerm = term
	}

	published, err := s.searchPublished(ctx, searchTerm)
	if err != nil {
		return nil, err
	}

	if len(published) == 0 && brandShort != "" && !strings.EqualFold(strings.TrimSpace(brandShort), searchTerm) {
		published, err = s.searchPublished(ctx, strings.TrimSpace(brandShort))
		if err != nil {
			return nil, err
		}
	}

	queryKw := eventKeywords(eventName)
	for w := range eventKeywords(brandShort) {
		queryKw[w] = true
	}

	scored := make([]scoredEmail, 0, len(published))
	for _, e := range published {
		score := 0
		if len(queryKw) > 0 {
			nameKw := eventKeywords(e.Name)
			for w := range queryKw {
				if nameKw[w] {
					score++
				}
			}
		}
		scored = append(scored, scoredEmail{score: score, email: e})
	}
	if len(queryKw) > 0 {
		matched := make([]scoredEmail, 0, len(scored))
		for _, p := range scored {
			if p.score > 0 {
				matched = append(matched, p)
			}
		}
		if len(matched) > 0 {
			scored = matched
		}
	}
	// published is already sorted -publishDate server-side; a stable sort
	// on keyword score alone preserves that as the tiebreak.
	sort.SliceStable(scored, func(i, j int) bool { return scored[i].score > scored[j].score })

	if len(scored) > limit {
		scored = scored[:limit]
	}

	cache := map[string]model.ListBrief{}
	results := make([]model.LastSentEmail, 0, len(scored))
	for _, p := range scored {
		e := p.email
		includedIDs := append(append([]string{}, e.To.ContactLists.Include...), e.To.ContactIlsLists.Include...)
		excludedIDs := append(append([]string{}, e.To.ContactLists.Exclude...), e.To.ContactIlsLists.Exclude...)

		includedBriefs := make([]model.ListBrief, 0, len(includedIDs))
		for _, id := range includedIDs {
			includedBriefs = append(includedBriefs, s.resolveListBrief(ctx, id, cache))
		}
		excludedBriefs := make([]model.ListBrief, 0, len(excludedIDs))
		for _, id := range excludedIDs {
			excludedBriefs = append(excludedBriefs, s.resolveListBrief(ctx, id, cache))
		}

		sentAt := e.PublishDate
		if sentAt == "" {
			sentAt = e.UpdatedAt
		}
		hubspotURL := ""
		if e.ID != "" {
			hubspotURL = fmt.Sprintf("https://app.hubspot.com/email/%s/edit/%s", s.portalID, e.ID)
		}

		results = append(results, model.LastSentEmail{
			EmailID:          e.ID,
			EmailName:        e.Name,
			SentAt:           sentAt,
			HubSpotURL:       hubspotURL,
			IncludedLists:    dedupeBriefs(includedBriefs),
			SuppressionLists: dedupeBriefs(excludedBriefs),
		})
	}
	return results, nil
}

func (s *LastSentService) searchPublished(ctx context.Context, term string) ([]model.EmailSummary, error) {
	all, err := s.emails.SearchEmails(ctx, domain.EmailSearchOptions{NameContains: term, Limit: 30, OrderBy: "-publishDate"})
	if err != nil {
		return nil, err
	}
	published := make([]model.EmailSummary, 0, len(all))
	for _, e := range all {
		if e.State == "PUBLISHED" {
			published = append(published, e)
		}
	}
	return published, nil
}
